### PR TITLE
feat: use virtual title in notifications if title is not supported

### PR DESCRIPTION
**Pull request by Akmadan23**
Having `[NvimTree]` as a notification header while using a notification system that actually supports title is redundant, so with this simple change it will be added only if needed.

### CODE REVIEW

**Comment by Akmadan23** [comment]
And maybe now it's time to raise stylua's column width, splitting these boolean expressions would be really confusing.

**Comment by alex-courtis** [comment]
It is hard to determine whether we have an alternate notify plugin installed.

Could we be clever and look for the `__call` metamethod on `vim.inspect`?

Experimenting:
```
[2023-10-02 10:45:23] [dev] default vim.notify=<function 1>

[2023-10-02 10:49:32] [dev] "rcarriga/nvim-notify" vim.notify={
  _config = <function 1>,
  _print_history = <function 2>,
  async = <function 3>,
  dismiss = <function 4>,
  history = <function 5>,
  instance = <function 6>,
  notify = <function 7>,
  open = <function 8>,
  pending = <function 9>,
  setup = <function 10>,
  <metatable> = {
    __call = <function 11>
  }
}

[2023-10-02 10:55:12] [dev] "folke/noice.nvim" vim.notify=<function 1>
```

That's a no...

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This will result in user happiness, thanks for taking it on.

We have options:
1. Harden detection. It might be nicer to put it in `notify.setup` so it runs at a controlled time rather than on module load.
2. Add `notify.prefix` boolean|string or similar, default `"[NvimTree]\n"`

I prefer 1 - we have too many options and users can add additional notify plugin detection if needed.

**Comment by alex-courtis** [comment]
> And maybe now it's time to raise stylua's column width, splitting these boolean expressions would be really confusing.

Yeah, this is getting silly. The intent is to keep statements minimal however sometimes you just have to have long lines.

```
      if
        (filters.config.filter_no_buffer or renderer.config.highlight_opened_files ~= "none")
        and vim.bo[data.buf].buftype == ""
      then
```

How about trying, say, 140 on a new branch and we see how it looks?

**Comment by alex-courtis** [comment]
Would it be possible for us to move `notify_once` to `notify.lua`?

File: lua/nvim-tree/notify.lua [hunk 1]
```diff
@@ -1,5 +1,9 @@
 local M = {}
 
+M.supports_title = (pcall(require, "notify") and vim.notify == require "notify" or vim.notify == require("notify").notify)
+  or (pcall(require, "noice") and vim.notify == require("noice").notify or vim.notify == require("noice.source.notify").notify)
```
[the remark this answers is not in the document]
Comment (Akmadan23) [reply lua/nvim-tree/notify.lua:4]: Right, I forgot to put parentheses on the second part, it did not show up to me because I have nvim-notify enabled.

**Comment by Akmadan23** [comment]
> We have options:
> 
>   1. Harden detection. ~It might be nicer to put it in `notify.setup` so it runs at a controlled time rather than on module load.~
> 
>   2. Add `notify.prefix` boolean|string or similar, default `"[NvimTree]\n"`
> 
> 
> I prefer 1 - we have too many options and users can add additional notify plugin detection if needed.

Yeah, I also initially considered to add a user option to manage it, but this is such a slight and almost irrelevant improvement that probably no one would care about going out of their way to change it.

In order to harden detection, as it stands now the notification plugin must be loaded before nvim-tree, but that's something that the user should do manually, so it isn't ideal...

Otherwise we can put the boolean expression inside a function that gets evaluated each time. I'm trying to create an hybrid that evaluates the expression only once (the first time it gets called) and then reuse its result without other pointless evaluations.

**Comment by Akmadan23** [comment]
> Would it be possible for us to move `notify_once` to `notify.lua`?

That would make sense, but there are a couple of choices to be made to make it consistent with other members of the `notify` module. Maybe we can discuss it in a future PR (unless it can be useful to achieve our goal here).

**Comment by Akmadan23** [comment]
> How about trying, say, 140 on a new branch and we see how it looks?

140 seems reasonable, I'm not sure what you mean by seeing how it looks on a new branch however. I mean, if we want to merge this when it will be ready I don't see any other decent alternative...

**Comment by Akmadan23** [comment]
This is the most efficient and effective solution I can find

**Comment by alex-courtis** [comment]
> Edit: now that I think about it... Can't we just use `notify.warn`? The "once" part seems unnecessary since in both cases the notification will be launched once.

Thank you. That simplifies greatly.

**Review by alex-courtis — APPROVED** [review]
- [x] nvim-notify
- [x] noice
- [x] notifier

**Comment by alex-courtis** [comment]
> 140 seems reasonable, I'm not sure what you mean by seeing how it looks on a new branch however. I mean, if we want to merge this when it will be ready I don't see any other decent alternative...

Added a temporary suppression.

Changing format disrupts PRs in flight, so we will do that when the queue is empty.

**Pull request merged**